Mathlib/RingTheory/AdicCompletion/Basic.lean

```diff
@@ -201,67 +203,259 @@ instance (priority := 100) of_subsingleton [Subsingleton M] : IsPrecomplete I M
     ((eval I M n).map_smul r x).symm ▸
       Quotient.inductionOn' (eval I M n x) fun x => SModEq.zero.2 <| smul_mem_smul hr mem_top)
     fun _ _ ih1 ih2 => by rw [LinearMap.map_add, ih1, ih2, LinearMap.map_zero, add_zero]⟩
 
-end adicCompletion
+@[simp]
+theorem transitionMap_mk {m n : ℕ} (hmn : m ≤ n) (x : M) :
+    transitionMap I M hmn
+      (Submodule.Quotient.mk (p := (I ^ n • ⊤ : Submodule R M)) x) =
+      Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) x := by
+  rfl
+
+@[simp]
+theorem transitionMap_eq (n : ℕ) : transitionMap I M (Nat.le_refl n) = LinearMap.id := by
+  ext
+  simp
+
+@[simp]
+theorem transitionMap_comp {m n k : ℕ} (hmn : m ≤ n) (hnk : n ≤ k) :
+    transitionMap I M hmn ∘ₗ transitionMap I M hnk = transitionMap I M (hmn.trans hnk) := by
+  ext
+  simp
+
+@[simp]
+theorem transitionMap_comp_apply {m n k : ℕ} (hmn : m ≤ n) (hnk : n ≤ k)
+    (x : M ⧸ (I ^ k • ⊤ : Submodule R M)) :
+    transitionMap I M hmn (transitionMap I M hnk x) = transitionMap I M (hmn.trans hnk) x := by
+  change (transitionMap I M hmn ∘ₗ transitionMap I M hnk) x = transitionMap I M (hmn.trans hnk) x
+  simp
+
+@[simp]
+theorem transitionMap_comp_eval_apply {m n : ℕ} (hmn : m ≤ n) (x : AdicCompletion I M) :
+    transitionMap I M hmn (x.val n) = x.val m :=
+  x.property hmn
+
+@[simp]
+theorem transitionMap_comp_eval {m n : ℕ} (hmn : m ≤ n) :
+    transitionMap I M hmn ∘ₗ eval I M n = eval I M m := by
+  ext x
+  simp
+
+/-- A sequence `ℕ → M` is an `I`-adic Cauchy sequence if for every `m ≤ n`,
+`f m ≡ f n` modulo `I ^ m • ⊤`. -/
+def IsAdicCauchy (f : ℕ → M) : Prop :=
+  ∀ {m n}, m ≤ n → f m ≡ f n [SMOD (I ^ m • ⊤ : Submodule R M)]
+
+/-- The type of `I`-adic Cauchy sequences. -/
+def AdicCauchySequence : Type _ := { f : ℕ → M // IsAdicCauchy I M f }
+
+namespace AdicCauchySequence
+
+/-- The type of `I`-adic cauchy sequences is a submodule of the product `ℕ → M`. -/
+def submodule : Submodule R (ℕ → M) where
+  carrier := { f | IsAdicCauchy I M f }
+  add_mem' := by
+    intro f g hf hg m n hmn
+    exact SModEq.add (hf hmn) (hg hmn)
+  zero_mem' := by
+    intro _ _ _
+    rfl
+  smul_mem' := by
+    intro r f hf m n hmn
+    exact SModEq.smul (hf hmn) r
+
+instance : CoeFun (AdicCauchySequence I M) (fun _ ↦ ℕ → M) where
+  coe f := f.val
+
+instance : AddCommGroup (AdicCauchySequence I M) :=
+  inferInstanceAs <| AddCommGroup (AdicCauchySequence.submodule I M)
+
+instance : Module R (AdicCauchySequence I M) :=
+  inferInstanceAs <| Module R (AdicCauchySequence.submodule I M)
+
+@[simp]
+theorem val_add (n : ℕ) (f g : AdicCauchySequence I M) : (f + g) n = f n + g n :=
+  rfl
+
+@[simp]
+theorem val_sub (n : ℕ) (f g : AdicCauchySequence I M) : (f - g) n = f n - g n :=
+  rfl
+
+@[simp]
+theorem val_smul (n : ℕ) (r : R) (f : AdicCauchySequence I M) : (r • f) n = r • f n :=
+  rfl
+
+@[simp]
+theorem val_zero (n : ℕ) : (0 : AdicCauchySequence I M) n = 0 :=
+  rfl
+
```

Comment:
Maybe a simp lemma `f.1 x = f x` if the latter is simp NF.

Reply:
This is proven by `simp`, should I add it anyway?

Reply:
Oh maybe coefun is reducible or something similar.
I think the lemma might still be useful though?

Reply:
CI says "LHS equals RHS syntactically", so I removed it again.

Reply:
Ah coefun is smarter than I think. Sorry for the confusion.